solutions/usaco-923.mdx

```diff
@@ -0,0 +1,420 @@
+	for i in arr:
+		curr_sum = max(curr_sum + i, 0)
+		best_sum = max(best_sum, curr_sum)
+	return best_sum
+```
+
+</PySection>
+</LanguageSection>
+
+We can extend this to the second dimension
+(albeit with an extra $N^2$ factor in the complexity)
+by iterating through all ranges of columns in the matrix and running a 1D Kadane's
+where the sum of each element is the subarray is the sum of the column range of a row.
+
+For example, if our current column range was $[0,2]$, then we'd run our 1D Kadane's
+on the following array:
+$$
+[0,2,0,0,1,1,1,0,0,0]
+$$
+
+### Solving the Second Matrix
+
+The key to getting the max with two submatrices is noticing that there will
+always be a line, vertical or horizontal, that can divide the grid such that
+one rectangle is in one section and the other rectangle is in the other section.
+
+Building on this observation, we can go through all the lines we can draw
+through the matrix and get the max subarray on both sides.
+After this, we can take the max of the results of these lines and add it
+to the number of cells that were initially painted with the optimal number
+of layers of paint to get our final answer.
+
+We need a method to efficiently calculate the max submatrices for all the split
+regions, though.
+To start off, let's define four arrays which will have the following
+***initial*** values:
+
+* $\texttt{top\_best[i]}$: the best matrix whose ***lower*** side is the $i$th row.
+* $\texttt{bottom\_best[i]}$: the best matrix whose ***upper*** side is the $i$th row.
+* $\texttt{left\_best[i]}$: the best matrix whose ***right*** side is the $i$th column.
+* $\texttt{right\_best[i]}$: the best matrix whose ***left*** side is the $i$th column.
+
+We fill these arrays by running a 2D Kadane's starting from the top, bottom,
+left, or right, and updating the corresponding arrays with the max submatrix value.
+
+Next, we do a cumulative maximum from the start or from the end depending on the array
+to get the arrays with the following updated values:
+
+* $\texttt{top\_best[i]}$: the best matrix that lies ***at or above*** the $i$th row.
+* $\texttt{bottom\_best[i]}$: the best matrix that lies ***at or below*** $i$th row.
+* $\texttt{left\_best[i]}$: the best matrix that's ***at or to the left of*** the $i$th column.
+* $\texttt{right\_best[i]}$: the best matrix that's ***at or to the right of*** the $i$th column.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N^3)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <algorithm>
+
+using std::cout;
+using std::endl;
+using std::vector;
+using std::max;
+
+const int WIDTH = 200;
+
+int main() {
+	std::ifstream read("paintbarn.in");
+	int rect_num;
+	int optimal_amt;
+	read >> rect_num >> optimal_amt;
+
+	vector<vector<int>> barn(WIDTH, vector<int>(WIDTH));
+	for (int r = 0; r < rect_num; r++) {
+		int x1, y1, x2, y2;
+		read >> x1 >> y1 >> x2 >> y2;
+		for (int y = y1; y < y2; y++) {
```

Comment:
this is O(N*y), can you change it to O(N+xy) like the editorial (http://www.usaco.org/current/data/sol_paintbarn_gold_feb19.html)?

Reply:
i mean it still runs fine? don't think it has any big impact on readability either

Reply:
sure